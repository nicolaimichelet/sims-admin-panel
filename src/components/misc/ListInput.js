import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { FlatButton } from 'material-ui';
import RaisedButton from 'material-ui/RaisedButton/RaisedButton';
import IconButton from 'material-ui/IconButton';

import {red700, white} from 'material-ui/styles/colors'


export class ListInput extends Component{
  constructor(props){
    super(props);
    let values = props.values || [];
    this.state = {
      values: values,
      count: props.count || props.min || values.length 
    }
  }

  onInputChange(idx, value){
    let values = this.state.values.slice();
    values[idx] = value;
    
    this.setState({
      values: values
    }, () => {
      this.props.onChange(this.state.values);
    });
  }


  isControlled(){
    return this.props.count != null || this.props.values != null; 
  }

  onCountChange(count){
    let min = this.props.min || 0;
    let max = this.props.max || Infinity;
    this.setState({
      count: Math.min(Math.max(count, min), max)
    }, () => {
      this._updateList(this.props);
    });
  }


  removeField(idx){
    let values = this.state.values.slice();
    values.splice(idx, 1);
    this.setState({
      values: values
    }, () => {
      this.onCountChange(this.state.count - 1);
    })
  }

  onComponentDidMount(){
    this._updateList(this.props);
  }

  componentWillReceiveProps(nextProps){
    this._updateList(nextProps);
  }

  _updateList(props){
    this.setState({
      count: (this.props.count != props.count) ? props.count : this.state.count,
      values: (this.props.values != props.values) ? props.values : this.state.values
    }, () => {
      let values = this.state.values.slice(0, this.state.count);
      this.setState({
        values: values
      });
    });
  }





  render(){

    let fields = [];
    let FieldType = this.props.component;
    for(let i = 0; i < this.state.count; i++){
      fields.push(
        <li key={i}>
          <FieldType key={i} value={this.state.values[i]} onChange={(...a) => {this.onInputChange(i, ...a)}}/>
          <IconButton iconStyle={{color: red700}} iconClassName="material-icons" onClick={() => this.removeField(i)}>remove_circle</IconButton>
        </li>
      );
    }

    return (
      <div>
        <ul style={{ listStyleType: "none"}}>
          {fields}
        </ul>
        <RaisedButton label={'New'} labelColor={white} primary={true} onClick={() => this.onCountChange(this.state.count + 1)}/>
      </div>
    );
  }
}


ListInput.defaultProps = {
  min: 0,
  max: Infinity,
  component: () => <p>"none"</p>
}

ListInput.propTypes = {
  count: PropTypes.number,
  max: PropTypes.number,
  min: PropTypes.number,
  values: PropTypes.array
}

