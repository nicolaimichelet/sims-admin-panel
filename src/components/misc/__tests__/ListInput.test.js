import React from 'react';
import ReactDOM from 'react-dom';
import { ListInput } from 'components/misc/ListInput.js';

import { mount, shallow } from 'enzyme';



describe('ListInput', () => {
  
  it('renders with no props', () => {
    const wrapper = shallow(<ListInput />);
    expect(wrapper).toMatchSnapshot();
  });
});
