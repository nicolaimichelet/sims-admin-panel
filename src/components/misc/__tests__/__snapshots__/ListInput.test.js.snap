// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListInput renders with no props 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ListInput
    component={[Function]}
    max={Infinity}
    min={0}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <ul
          style={
                    Object {
                              "listStyleType": "none",
                            }
          }
>
          
</ul>,
        <RaisedButton
          disabled={false}
          fullWidth={false}
          label="New"
          labelColor="#ffffff"
          labelPosition="after"
          onClick={[Function]}
          primary={true}
          secondary={false}
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [],
          "style": Object {
            "listStyleType": "none",
          },
        },
        "ref": null,
        "rendered": Array [],
        "type": "ul",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "disabled": false,
          "fullWidth": false,
          "label": "New",
          "labelColor": "#ffffff",
          "labelPosition": "after",
          "onClick": [Function],
          "primary": true,
          "secondary": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <ul
            style={
                        Object {
                                    "listStyleType": "none",
                                  }
            }
>
            
</ul>,
          <RaisedButton
            disabled={false}
            fullWidth={false}
            label="New"
            labelColor="#ffffff"
            labelPosition="after"
            onClick={[Function]}
            primary={true}
            secondary={false}
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "style": Object {
              "listStyleType": "none",
            },
          },
          "ref": null,
          "rendered": Array [],
          "type": "ul",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "disabled": false,
            "fullWidth": false,
            "label": "New",
            "labelColor": "#ffffff",
            "labelPosition": "after",
            "onClick": [Function],
            "primary": true,
            "secondary": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
